pkg/tun/tcp/handler.go

```diff
@@ -313,14 +313,17 @@ func (h *handler) sendSyn(ctx context.Context, ackNumber uint32, ack bool) {
 // writeToTunLoop sends the packages read from the fromMgr channel to the TUN device
 func (h *handler) writeToTunLoop(ctx context.Context) {
 	defer close(h.readyToFin)
-	ticker := time.NewTicker(100 * time.Millisecond)
 	for {
 		var cm connpool.Message
 		select {
+		// FIXME(josecv): Prioritize draining the fromMgr queue over terminating; there's a subtle race
```

Comment:
If this actually makes a difference, then I think that is because the timer is increased from 100 to 500 ms. No prioritization is made because Go will shuffle the cases in a channel select on each call to ensure that order is insignificant.

I think we need a proper drain timeout. E.g. have a separate goroutine that only deals with the `h.fromMgr` channel and another that deals with `context.Done()` The latter will then close the `h.fromMgr` after some configurable delay. Perhaps that channel could also be closed at the writer end when the peer sends a close control message.

Reply:
> No prioritization is made because Go will shuffle the cases in a channel select on each call to ensure that order is insignificant.

oh wow TIL. It must be the timer change then that's fixed this, because I have not seen the tests fail on this branch.

> I think we need a proper drain timeout. E.g. have a separate goroutine that only deals with the h.fromMgr channel and another that deals with context.Done() The latter will then close the h.fromMgr after some configurable delay. Perhaps that channel could also be closed at the writer end when the peer sends a close control message.

Yeah I think that makes sense, my only worry about closing the channel is that we need to make sure the other end's context is cancelled too, or we'll get a panic when it tries to write to the channel. Right now I don't think that's the case, as I've seen the other end hang on a `fromMgr <- msg` operation after this end's context is finished.